snapshots+code-caching.md

```diff
@@ -0,0 +1,84 @@
+# Snapshots and Code Caching
+
+This document explains techniques used by V8 in order to avoid having to re-compile and
+optimized JavaScript whenever an application that embeds it (i.e. Chrome or Node.js) starts up
+fresh.
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*
+
+- [Code Caching](#code-caching)
+  - [Chrome's Use of Code Caching](#chromes-use-of-code-caching)
+  - [Resources](#resources)
+- [Startup Snapshots](#startup-snapshots)
+  - [Custom Startup Snapshots](#custom-startup-snapshots)
+  - [Lazy Deserialization](#lazy-deserialization)
+  - [Resources](#resources-1)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Code Caching
+
+- lessens overhead of parsing + compiling script
+- uses cached data to recreate previous compilation result
+- exposed via V8's API to embedders
+  - pass `v8::ScriptCompiler::kProduceCodeCache` as an option when compiling script
+  - cached data is attached to source object to be retrieved via
+    `v8::ScriptCompiler::Source::GetCachedData`
+  - can be persisted for later
+  - later cache data can be attached to the source object and passed
+    `v8::ScriptCompiler::kConsumeCodeCache` as an option to cause V8 to bypass compileing the
+    code and deserialize the provided cache data instead
+- V8 6.6 caches top level code as well as code generated _after_ script's top-level execution,
+  which means that lazily compiled functions are included in the cache
+
+### Chrome's Use of Code Caching
+
+Since Chrome embeds V8 it can make use of Code Caching and does so as follows.
+
+- cold load: page loaded for the first time and thus no cached data is available
+- warm load: page loaded before and caches compiled code along with the script file in disk
```

Comment:
Current heurisric also includes that the last load needs to be within the last 72 hours.

Reply:
Wow, I had no idea about that part .. I'll add that.
Is there a resource somewhere about that detail? Otherwise I'll try to find something related in the v8 code.

Reply:
This is actually [here](https://cs.chromium.org/chromium/src/third_party/blink/renderer/bindings/core/v8/v8_script_runner.cc?l=269&rcl=c59618d0f92b57e4dcfb903f3c99bb0574eac340). We also require the script source to be larger than 1k.

Reply:
Addresed in b11e195